File: algorithms/quick_sort.go
package main

import "fmt"

func QuickSort(nums []int) {
	if nums == nil || len(nums) <= 1 {
		return
	}
	var pos int = partition(nums)
	QuickSort(nums[:pos])
	QuickSort(nums[pos+1:])
}

func partition(nums []int) int {
	pivot := nums[0]
	left, right := 1, len(nums)-1

	for left <= right {
		for left <= right && nums[left] <= pivot {
			left++
		}
		for left <= right && nums[right] > pivot {
			right--
		}
		if left > right {
			break
		}
		nums[left], nums[right] = nums[right], nums[left]
	}
	nums[0], nums[right] = nums[right], nums[0]
	return right
}

func main() {
	var nums = make([]int, 0, 8)
	nums = []int{2, 8, 4, 7, 6, 1, 5, 3}

	fmt.Println(nums)
	QuickSort(nums)
	fmt.Println(nums)
}

File: algorithms/quick_sort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


def quick_sort(nums, start, end):
    if start > end:
        return
    pos = partition(nums, start, end)
    quick_sort(nums, start, pos-1)
    quick_sort(nums, pos+1, end)


def partition(nums, left, right):
    pivot = nums[left]
    start = left
    left += 1
    while 1:
        while left <= right and nums[left] <= pivot:
            left += 1
        while left <= right and nums[right] > pivot:
            right -= 1
        if left > right:
            break
        else:
            nums[left], nums[right] = nums[right], nums[left]
    nums[start], nums[right] = nums[right], nums[start]
    return right


if __name__ == '__main__':
    import random
    nums = range(20)
    random.shuffle(nums)
    print nums
    quick_sort(nums, 0, len(nums) - 1)
    print nums
